src/pipeline.rs

```diff
@@ -60,6 +60,23 @@ impl Sink for Pipeline {
         debug_assert!(self.enqueued.is_empty());
         self.inner.poll_complete()
     }
+
+    fn close(&mut self) -> Poll<(), Self::SinkError> {
+        self.poll_complete()?;
+        self.inner.close()
+    }
+}
+
+impl Drop for Pipeline {
+    fn drop(&mut self) {
+        loop {
+            match self.close() {
+                Ok(Async::Ready(())) => break,
+                Ok(Async::NotReady) => {}
```

Comment:
This could turn into a pretty tight loop if there's something `inner` is waiting on before closing. That plus the error handling question below makes me think we'd be better off requiring the caller to use `close` than doing this through `Drop`.

Reply:
:+1: I'm happy to just call `close()` in this one particular case, but do you have thoughts on ensuring that we call `close()` everywhere `forward` or `send_all` is used?

Reply:
(my concern is that we'll forget about this again)

Reply:
We will need to move away from the futures 0.1 `send_all`/`forward` methods before too long, so we could add our own method to `Pipeline` directly that makes sure to flush before returning an error.